lfsapi/client.go

```diff
@@ -0,0 +1,149 @@
+package lfsapi
+
+import (
+	"crypto/tls"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/git-lfs/git-lfs/errors"
+	"github.com/rubyist/tracerx"
+)
+
+func (c *Client) Do(req *http.Request) (*http.Response, error) {
+	res, err := c.doWithRedirects(c.httpClient(req.Host), req, nil)
+	if err != nil {
+		return res, err
+	}
+
+	return res, c.handleResponse(res)
+}
+
+func (c *Client) doWithRedirects(cli *http.Client, req *http.Request, via []*http.Request) (*http.Response, error) {
+	if seeker, ok := req.Body.(io.Seeker); ok {
+		seeker.Seek(0, io.SeekStart)
+	}
+
+	res, err := cli.Do(req)
+	if err != nil {
+		return res, err
+	}
+
+	if res.StatusCode != 307 {
```

Comment:
One of our roadmap items is to respond to 301s as well. Do you think we could add an `|| res.StatusCode == 301`, and maintain a cache of permanent redirections? I'm thinking something like a `map[*url.URL]*url.URL` map of from -> to.